langserver/build_context.go

```diff
@@ -13,6 +13,40 @@ import (
 	"golang.org/x/net/context"
 )
 
+// withCancelContext creates a build.Context which wraps the input
+// *build.Context and aborts pending operations after ctx.Done
+func (h *LangHandler) withCancelContext(ctx context.Context, w *build.Context) *build.Context {
+	// We're mutating the build context that we intend to wrap, so copy it.
+	copy := *w
+	bctx := &copy
```

Comment:
you can avoid the copy since we always use `withCancelContext`. I'd just inline in `BuildContext` and document there that if the context is cancelled we fail filesystem operations.